osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObjectContainer.cs

```diff
@@ -0,0 +1,334 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Performance;
+using osu.Framework.Graphics.Pooling;
+using osu.Game.Rulesets.Catch.UI;
+using osu.Game.Utils;
+using osuTK;
+
+namespace osu.Game.Rulesets.Catch.Objects.Drawables
+{
+    public class CaughtObjectContainer : CompositeDrawable
+    {
+        public Container<DrawableCaughtObject> StackedObjectContainer { get; }
+
+        private readonly Container<DrawableCaughtObject> droppedObjectTarget;
+
+        private readonly LifetimeEntryManager lifetimeManager = new LifetimeEntryManager();
+
+        private readonly Dictionary<LifetimeEntry, DrawableCaughtObject> drawableMap = new Dictionary<LifetimeEntry, DrawableCaughtObject>();
+
+        private readonly HashSet<StackedObjectEntry> aliveStackedObjects = new HashSet<StackedObjectEntry>();
+
+        private readonly DrawablePool<DrawableCaughtFruit> caughtFruitPool;
+        private readonly DrawablePool<DrawableCaughtBanana> caughtBananaPool;
+        private readonly DrawablePool<DrawableCaughtDroplet> caughtDropletPool;
+
+        /// <summary>
+        /// The randomness used to compute position in stack.
+        /// It is incremented for each <see cref="AddStackObject"/> or <see cref="AddDropObject"/> call and decremented for each <see cref="RemoveEntry"/> call to make a replay consistent.
+        /// </summary>
+        private int randomSeed = 1;
+
+        public CaughtObjectContainer(Container<DrawableCaughtObject> droppedObjectTarget)
+        {
+            this.droppedObjectTarget = droppedObjectTarget;
+
+            InternalChildren = new Drawable[]
+            {
+                caughtFruitPool = new DrawablePool<DrawableCaughtFruit>(50),
+                caughtBananaPool = new DrawablePool<DrawableCaughtBanana>(100),
+                // less capacity is needed compared to fruit because droplet is not stacked
+                caughtDropletPool = new DrawablePool<DrawableCaughtDroplet>(25),
+                StackedObjectContainer = new Container<DrawableCaughtObject>()
+            };
+
+            lifetimeManager.EntryBecameAlive += entryBecameAlive;
+            lifetimeManager.EntryBecameDead += entryBecameDead;
+        }
+
+        /// <summary>
+        /// Add a caught object to the stack.
+        /// </summary>
+        public CaughtObjectEntry AddStackObject(IHasCatchObjectState source, Vector2 positionInStack)
+        {
+            var delayedDropEntry = new DroppedObjectEntry(positionInStack, source);
+            var stackEntry = new StackedObjectEntry(positionInStack, delayedDropEntry, source);
+
+            addImmediateEntry(stackEntry);
+
+            // `DropStackedObjects` may be called before lifetime update.
+            if (stackEntry.LifetimeStart <= Clock.CurrentTime)
```

Comment:
I don't know if it's just me but I'd prefer this conditional was flipped, as it feels more natural for me to have the instant being checked on the left side of the comparison against a start time of an interval (was part of the reason why I misread earlier and thought this was buggy).